Add C source code inspection tab for TParseNode

hernanmd opened the pull request:
This PR add an extension method to VMMaker-Tools to access a new tab with the C code translation of any inspected TParseNode. See the following screenshots:

<img width="870" alt="Screenshot 2023-09-06 at 21 03 26" src="https://github.com/pharo-project/pharo-vm/assets/4825959/2467549a-5a38-4fcd-8575-f1cdaee14532">
<img width="1051" alt="Screenshot 2023-09-06 at 21 05 43" src="https://github.com/pharo-project/pharo-vm/assets/4825959/dfebaa59-d1bc-482c-852b-b0c985f1b2fe">

To test it inspect the follwing nodes:

```smalltalk
TStatementListNode new
		              setArguments: #(  )
		              statements: { 
				              (TSendNode new
					               setSelector: #foo
					               receiver: (TConstantNode value: 1)
					               arguments: { (TConstantNode value: 2) }).
				              (TSendNode new
					               setSelector: #foo
					               receiver: (TConstantNode value: 3)
					               arguments: { (TConstantNode value: 4) }).
				              (TSendNode new
					               setSelector: #foo
					               receiver: (TConstantNode value: 5)
					               arguments: { (TConstantNode value: 6) }) }.
```

Commit:
Add an extension method to VMMaker-Tools to access a new tab with the C code translation of any inspected TParseNode.

guillep commented on smalltalksrc/VMMaker-Tools/TParseNode.extension.st:
```diff
@@ -1,5 +1,26 @@
+					definingClass: self class;
+					yourself).
+			cg pushScope: TStatementListNode new.
+			(self respondsTo: #expression)
```
I don't think this is correct. Not all nodes answering to `#expression` are expressions.
For example, return nodes contain an expression (the expression to return) but are not expressions themselves!

Maybe we miss an `#isExpression` message?

Also, it would be valuable to see both translations (as expression and not), maybe in two tabs?
There is a way to make inspector tabs appear conditionally I think.

guillep replied:
Yes, you need to define a method with the same name receiving an argument. For example, check 

```smalltalk
AbstractFileReference >> inspectionContents
	<inspectorPresentationOrder: 0 title: 'Contents'> 
	...

AbstractFileReference >> inspectionContentsContext: aContext 
	aContext active: self isFile
```

We could have an equivalent with `isExpression`.

Then, if you ask yourself what nodes are expressions, you can check the ones redefining `asCASTExpressionIn:`

hernanmd replied:
I checked the active usage but I couldn't fully understood how it works, because somewhere I should set the a method selector, I tried doing something like 

```smalltalk
	self isExpression
		ifTrue: [ aContext methodSelector: #asCASTExpressionIn: ]
		ifFalse: [ aContext methodSelector: #asCASTIn: ].
```

or: 

```smalltalk
inspectNode: aContext 

	aContext active: self isExpression
```

But then if we have two tabs we wouldn't make use of the #isExpression right?

guillep replied:
It's by name convention. If you have a method defining an inspector tab that is called `myInspectorTab`, then you can define also `myInspectorTabContext: aContext` that wil have a configuration context as argument. And you can use that to set different properties to the tab (like visibility or other)

guillep replied:
The idea is to only show the C expression for those that make sense, those that are not expressions (and will fail with an exception) could be filtered out

guillep reviewed (changes requested):
Very nice!!!

Small comments on the way, but I think this could save a lot of time of debugging ^^

Commit:
Add #isExpression to TAssignmentNode, TSendNode, TStatementListNode and TSwitchStmtNode, which redefines #asCASTExpressionIn: method.
Add #isExpression to TParseNode which answers false.
Add another palette which also display the C-AST in addition to the C-AST Expression.

hernanmd commented:
I add the #isExpression here anyway with tests, but let me know if there are useful enough to keep them.

Commit:
Refactoring a bit
Add tests

guillep commented:
If you think they are useful we can keep them.

I'm personally not so fond of those kinds of tests because they give no much value to me (just testing a constant) and we have to maintain the test :)

Commit:
Remove unnecessary tests
Add conditional inspection tabs

hernanmd commented:
> If you think they are useful we can keep them.
> 
> I'm personally not so fond of those kinds of tests because they give no much value to me (just testing a constant) and we have to maintain the test :)

I see the point, so I removed them.
Also pushed a version with conditional display of tabs IIUC.

guillep reviewed (approved):
LGTM! Maybe could be useful/nice for @ivojawer
